### PR TITLE
[Fix]fix the bug that mix_results may be None

### PR DESCRIPTION
**Pull request by Czm369**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Fix the bug that mix_results may be None.

## Modification

Some transform may return None, so we need to make sure that mix_results is not None.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/datasets/dataset_wrappers.py [hunk 1]
```diff
@@ -401,13 +401,16 @@ def __getitem__(self, idx):
...
+                    if not isinstance(indexes, collections.abc.Sequence):
+                        indexes = [indexes]
+                    mix_results = [
+                        copy.deepcopy(self.dataset[index]) for index in indexes
+                    ]
+                    if None not in mix_results:
+                        results['mix_results'] = mix_results
+                        break
 
             results = transform(results)
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:415]: some transform will return None here

File: mmdet/datasets/dataset_wrappers.py [hunk 2]
```diff
@@ -401,13 +401,16 @@ def __getitem__(self, idx):
...
-                ]
-                results['mix_results'] = mix_results
+                while True:
+                    indexes = transform.get_indexes(self.dataset)
+                    if not isinstance(indexes, collections.abc.Sequence):
+                        indexes = [indexes]
+                    mix_results = [
+                        copy.deepcopy(self.dataset[index]) for index in indexes
+                    ]
+                    if None not in mix_results:
```
Comment (ZwwWayne) [inline mmdet/datasets/dataset_wrappers.py:411]: It seems that the transform will re-generate all samples even if there is only one None in mix_results, we may consider make it more efficient.

File: mmdet/datasets/dataset_wrappers.py [hunk 3]
```diff
@@ -401,15 +401,26 @@ def __getitem__(self, idx):
...
-                if not isinstance(indexes, collections.abc.Sequence):
-                    indexes = [indexes]
-                mix_results = [
-                    copy.deepcopy(self.dataset[index]) for index in indexes
-                ]
-                results['mix_results'] = mix_results
-
-            results = transform(results)
+                while True:
+                    # RandomCrop and Albu may return None.
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:405]: This comment seems incorrect. RandomCrop and Albu do not belong to the load pipeline
'' Make sure the results passed the pipeline of the original dataset is not None.” would be more accurate.

File: mmdet/datasets/dataset_wrappers.py [hunk 4]
```diff
@@ -401,15 +401,26 @@ def __getitem__(self, idx):
...
+                    mix_results = [
+                        copy.deepcopy(self.dataset[index]) for index in indexes
+                    ]
+                    if None not in mix_results:
+                        results['mix_results'] = mix_results
+                        break
+
+            while True:
+                # To confirm the results passed Mosaic, MixUp
+                # or CopyPaste is not None.
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:419]: "To confirm the results passed the pipeline of the wrapper is not None."

File: mmdet/datasets/dataset_wrappers.py [hunk 5]
```diff
@@ -401,15 +401,26 @@ def __getitem__(self, idx):
...
-                indexes = transform.get_indexes(self.dataset)
-                if not isinstance(indexes, collections.abc.Sequence):
-                    indexes = [indexes]
-                mix_results = [
-                    copy.deepcopy(self.dataset[index]) for index in indexes
-                ]
-                results['mix_results'] = mix_results
-
-            results = transform(results)
+                while True:
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:404]: Need to add a limitation to these two loops to avoid infinite loop.

Comment (RangiLyu) [reply mmdet/datasets/dataset_wrappers.py:404]: And raise some detailed error messages to tell users their data processing pipeline may be incorrect.

File: mmdet/datasets/dataset_wrappers.py [hunk 6]
```diff
@@ -401,15 +407,34 @@ def __getitem__(self, idx):
...
+                    mix_results = [
+                        copy.deepcopy(self.dataset[index]) for index in indexes
+                    ]
+                    if None not in mix_results:
+                        results['mix_results'] = mix_results
+                        break
+                else:
+                    raise RuntimeError(
+                        'the loading pipeline of the original dataset'
+                        ' always return None.')
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:425]:
```suggestion
                        'The loading pipeline of the original dataset'
                        ' always return None. Please check the correctness of the dataset and its pipeline.')
```

File: mmdet/datasets/dataset_wrappers.py [hunk 7]
```diff
@@ -401,15 +407,34 @@ def __getitem__(self, idx):
...
+            for i in range(self.max_iters):
+                # To confirm the results passed the training pipeline
+                # of the wrapper is not None.
+                updated_results = transform(copy.deepcopy(results))
+                if updated_results is not None:
+                    results = updated_results
+                    break
+            else:
+                raise RuntimeError(
+                    'the training pipeline of the dataset wrapper'
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:436]: same as above

File: mmdet/datasets/dataset_wrappers.py [hunk 8]
```diff
@@ -353,13 +353,18 @@ class MultiImageMixDataset:
             dynamically. Default to None. It is deprecated.
         skip_type_keys (list[str], optional): Sequence of type string to
             be skip pipeline. Default to None.
+        max_iters (int): The maximum number of iterations for calling a
+            transform. If the number of iterations is greater than
+            `max_iters`, but results is stillNone, then the iteration is
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:358]:
```suggestion
            `max_iters`, but results is still None, then the iteration is
```

File: mmdet/datasets/dataset_wrappers.py [hunk 9]
```diff
@@ -353,13 +353,18 @@ class MultiImageMixDataset:
             dynamically. Default to None. It is deprecated.
         skip_type_keys (list[str], optional): Sequence of type string to
             be skip pipeline. Default to None.
+        max_iters (int): The maximum number of iterations for calling a
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:356]:
```suggestion
        max_refetch (int): The maximum number of iterations for calling a
```

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**